src/Microsoft.AspNetCore.SignalR.Common/Protocol/HandshakeProtocol.cs

```diff
@@ -17,26 +20,39 @@ public static class HandshakeProtocol
     {
         private const string ProtocolPropertyName = "protocol";
         private const string ProtocolVersionPropertyName = "version";
+        private const string MinorVersionPropertyName = "minorVersion";
         private const string ErrorPropertyName = "error";
         private const string TypePropertyName = "type";
 
         /// <summary>
-        /// The serialized representation of a success handshake.
+        /// Previously used to cache the errorless return message.
+        /// Now, return messages need a minor version number, so multiple are cached in a dictionary.
         /// </summary>
-        public static ReadOnlyMemory<byte> SuccessHandshakeData;
+        [Obsolete("A successful handshake should include a minor version, but the value here does not.")]
+        public static ReadOnlyMemory<byte> SuccessHandshakeData = new byte[] { 123, 34, 109, 105, 110, 111, 114, 86, 101, 114, 115, 105, 111, 110, 34, 58, 48, 125, 30 }; 
 
-        static HandshakeProtocol()
+        private static ConcurrentDictionary<IHubProtocol, ReadOnlyMemory<byte>> _messageCache = new ConcurrentDictionary<IHubProtocol, ReadOnlyMemory<byte>>();
+
+        public static ReadOnlySpan<byte> GetSuccessfulHandshake(IHubProtocol protocol)
         {
-            var memoryBufferWriter = MemoryBufferWriter.Get();
-            try
-            {
-                WriteResponseMessage(HandshakeResponseMessage.Empty, memoryBufferWriter);
-                SuccessHandshakeData = memoryBufferWriter.ToArray();
-            }
-            finally
+            ReadOnlyMemory<byte> result;
+
+            if(!_messageCache.TryGetValue(protocol, out result))
             {
-                MemoryBufferWriter.Return(memoryBufferWriter);
+                var memoryBufferWriter = MemoryBufferWriter.Get();
+                try
+                {
+                    WriteResponseMessage(new HandshakeResponseMessage(protocol.GetMinorVersion()), memoryBufferWriter);
+                    _messageCache.TryAdd(protocol, memoryBufferWriter.ToArray());
+                    result = (ReadOnlyMemory<byte>)memoryBufferWriter.ToArray();
```

Comment:
Just call `ToArray` once and store it in a variable. In fact, you can just store it in `result` first and then use `result` to add to the `_messageCache`. `ToArray` has to allocate new memory to store the data, so we don't want to call it twice.

Also, you don't need an explicit cast to `ReadOnlyMemory<byte>`. The following compiles just file: `ReadOnlyMemory<byte> foo = memoryBufferWriter.ToArray()`